src/relay_pic/main.c

```diff
@@ -0,0 +1,153 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <xc.h>
+#include <pic16f1826.h>
+#include "mcc_generated_files/adc.h"
+#include "mcc_generated_files/mcc.h"
+
+typedef unsigned char uint8_t;
+typedef unsigned short uint16_t;
+
+#define POWER (1 << 2)
+#define SELECT (1 << 3)
+#define LIM1 (1 << 7)
+#define LIM2 (1 << 6)
+#define DIP1 (1 << 4)
+#define DIP2 (1)
+#define DIP3 (1 << 2)
+#define DIP4 (1 << 3)
+#define LED (1 << 5)
+#define CURR_SENSE1 (1 << 1)
+#define CURR_SENSE2 (1)
+
+uint16_t dipInputs;
+uint16_t adcResult;
+uint16_t i2cSlaveRecv; // Data received on i2c
+uint16_t i2cSlaveSend; // Data to be sent on i2c
+
+void i2cSlaveInit(uint16_t address) {
+    SSPSTAT = 0x80;
+    SSPADD = address;
+    SSPCON = 0x36;
+    SSPCON2 = 0x01;
+    TRISB1 = 1; // SDA
+    TRISB4 = 1; // SCL
+    GIE = 1;
+    PEIE = 1;
+    SSP1IF = 0;
+    SSP1IE = 1;
+}
+
+void interrupt i2cSlaveInt(void) { 
+    if (SSP1IF == 1) {
+        uint16_t temp;
+        SSPCONbits.CKP = 0;
+       
+        // If overflow or collision.
+        if ((SSPCONbits.SSPOV) || (SSPCONbits.WCOL)) {
+            temp = SSPBUF; // Read the previous value to clear the buffer
+            SSPCONbits.SSPOV = 0; // Clear the overflow flag
+            SSPCONbits.WCOL = 0;   // Clear the collision bit
+            SSPCONbits.CKP = 1;
+        }
+
+        // If last byte was Address + write
+        if (!SSPSTATbits.D_nA && !SSPSTATbits.R_nW) {
+            temp = SSPBUF;
+            while(!BF);
+            i2cSlaveRecv = SSPBUF;
+            SSPCONbits.CKP = 1;
+            SSPM3 = 0;
+       }
+        
+       // If last byte was Address + read
+       else if(!SSPSTATbits.D_nA && SSPSTATbits.R_nW) {
+           temp = SSPBUF;
+           BF = 0;
+           SSPBUF = i2cSlaveSend;
+           SSPCONbits.CKP = 1;
+           while(SSPSTATbits.BF);
+       }
+       
+       SSP1IF = 0;
+    } // if
+}
+
+//void readAnalogInputs() {
+//    for (int i = 0; i < 1000; i++) {}
+//    ADCON0 = 0x01; // Turn ADC on
+//    ADCON0 |= 1 << 1; // set b[1] "go" bit
+//    uint8_t doneBit;
+//    do { //wait for ADC to complete (go bit switches to 0 automatically when done)
+//        doneBit = ADCON0 & (1 << 1);
+//    } while (doneBit); //while go bit is on (AD conversion in progress)
+//
+//    adcResult = (ADRESH << 8) | ADRESL; //combine two 8bit values into a 16bit value
+//
+//    ADCON0 = 0x00; //Turn ADC off return;
+//}
+
+void setPower(bool out) { 
+    if (out) { LATA |= POWER;} 
+    else { LATA &= ~POWER; }
+}
+
+void setSelect(bool out) {
+    if (out) { LATA |= SELECT;} 
+    else { LATA &= ~SELECT; }
+}
+
+void setLim1(bool out) {
+    if (out) { LATA |= LIM1;} 
+    else { LATA &= ~LIM1; }
+}
+
+void setLim2(bool out) {
+    if (out) { LATA |= LIM2; } 
+    else { LATA &= ~LIM2; }
+}
+
+void setLed(bool out) {

```

Comment:
The way we deal with LEDs on other boards (which i think is more readable) is like so:
```
#define BLUE_LED_ON() (LATC4 = 0)
#define BLUE_LED_OFF() (LATC4 = 1)
```
Then you can just call these macros wherever